File: cli/go/cmd/prompd/config_test.go
package main

import (
	"os"
	"testing"
)

func TestLoadConfig(t *testing.T) {
	// Save original environment
	originalEnvs := map[string]string{
		"OPENAI_API_KEY":       os.Getenv("OPENAI_API_KEY"),
		"ANTHROPIC_API_KEY":    os.Getenv("ANTHROPIC_API_KEY"),
		"GROQ_API_KEY":         os.Getenv("GROQ_API_KEY"),
		"PROMPD_DEFAULT_PROVIDER": os.Getenv("PROMPD_DEFAULT_PROVIDER"),
	}
	
	// Clean environment
	for key := range originalEnvs {
		os.Unsetenv(key)
	}
	
	defer func() {
		// Restore original environment
		for key, value := range originalEnvs {
			if value != "" {
				os.Setenv(key, value)
			} else {
				os.Unsetenv(key)
			}
		}
	}()

	t.Run("Default configuration", func(t *testing.T) {
		config, err := LoadConfig()
		if err != nil {
			t.Fatalf("LoadConfig failed: %v", err)
		}
		
		if config.MaxRetries != 3 {
			t.Errorf("Expected MaxRetries 3, got %d", config.MaxRetries)
		}
		
		if config.Timeout != 30 {
			t.Errorf("Expected Timeout 30, got %d", config.Timeout)
		}
		
		if config.APIKeys == nil {
			t.Error("APIKeys should not be nil")
		}
		
		if config.CustomProviders == nil {
			t.Error("CustomProviders should not be nil")
		}
	})
	
	t.Run("Environment variables", func(t *testing.T) {
		// Set test environment variables
		os.Setenv("OPENAI_API_KEY", "test-openai-key")
		os.Setenv("ANTHROPIC_API_KEY", "test-anthropic-key")
		os.Setenv("GROQ_API_KEY", "test-groq-key")
		os.Setenv("PROMPD_DEFAULT_PROVIDER", "openai")
		
		// Reset global config to force reload
		globalConfig = nil
		
		config, err := LoadConfig()
		if err != nil {
			t.Fatalf("LoadConfig failed: %v", err)
		}
		
		if config.APIKeys["openai"] != "test-openai-key" {
			t.Errorf("Expected OpenAI key 'test-openai-key', got '%s'", config.APIKeys["openai"])
		}
		
		if config.APIKeys["anthropic"] != "test-anthropic-key" {
			t.Errorf("Expected Anthropic key 'test-anthropic-key', got '%s'", config.APIKeys["anthropic"])
		}
		
		if config.APIKeys["groq"] != "test-groq-key" {
			t.Errorf("Expected Groq key 'test-groq-key', got '%s'", config.APIKeys["groq"])
		}
		
		if config.DefaultProvider != "openai" {
			t.Errorf("Expected DefaultProvider 'openai', got '%s'", config.DefaultProvider)
		}
	})
}

func TestGetAPIKey(t *testing.T) {
	config := &Config{
		APIKeys: map[string]string{
			"openai":    "config-openai-key",
			"anthropic": "config-anthropic-key",
		},
		CustomProviders: map[string]CustomProvider{
			"custom1": {
				APIKey:  "custom1-key",
				BaseURL: "http://localhost:8080",
				Enabled: true,
			},
		},
	}

	tests := []struct {
		name     string
		provider string
		expected string
	}{
		{
			name:     "OpenAI from config",
			provider: "openai",
			expected: "config-openai-key",
		},
		{
			name:     "Anthropic from config",
			provider: "anthropic",
			expected: "config-anthropic-key",
		},
		{
			name:     "Custom provider",
			provider: "custom1",
			expected: "custom1-key",
		},
		{
			name:     "Unknown provider",
			provider: "unknown",
			expected: "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := config.GetAPIKey(tt.provider)
			if result != tt.expected {
				t.Errorf("Expected %q, got %q", tt.expected, result)
			}
		})
	}
}

func TestGetDefaultProvider(t *testing.T) {
	tests := []struct {
		name            string
		defaultProvider string
		expected        string
	}{
		{
			name:            "Configured provider",
			defaultProvider: "anthropic",
			expected:        "anthropic",
		},
		{
			name:            "Empty provider - fallback to openai",
			defaultProvider: "",
			expected:        "openai",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			config := &Config{DefaultProvider: tt.defaultProvider}
			result := config.GetDefaultProvider()
			if result != tt.expected {
				t.Errorf("Expected %q, got %q", tt.expected, result)
			}
		})
	}
}

func TestGetDefaultModel(t *testing.T) {
	tests := []struct {
		name     string
		provider string
		expected string
	}{
		{
			name:     "OpenAI default model",
			provider: "openai",
			expected: "gpt-4o-mini",
		},
		{
			name:     "Anthropic default model",
			provider: "anthropic",
			expected: "claude-3-5-haiku-20241022",
		},
		{
			name:     "Groq default model",
			provider: "groq",
			expected: "llama3-8b-8192",
		},
		{
			name:     "Unknown provider",
			provider: "unknown",
			expected: "default",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			config := &Config{}
			result := config.GetDefaultModel(tt.provider)
			if result != tt.expected {
				t.Errorf("Expected %q, got %q", tt.expected, result)
			}
		})
	}
}

func TestIsProviderConfigured(t *testing.T) {
	config := &Config{
		APIKeys: map[string]string{
			"openai": "test-key",
		},
		CustomProviders: map[string]CustomProvider{
			"custom1": {
				APIKey:  "custom-key",
				BaseURL: "http://localhost:8080",
				Enabled: true,
			},
			"custom2": {
				BaseURL: "http://localhost:8081",
				Enabled: true,
				// No API key
			},
		},
	}

	tests := []struct {
		name     string
		provider string
		expected bool
	}{
		{
			name:     "Configured OpenAI",
			provider: "openai",
			expected: true,
		},
		{
			name:     "Unconfigured Anthropic",
			provider: "anthropic",
			expected: false,
		},
		{
			name:     "Ollama (local, no key needed)",
			provider: "ollama",
			expected: true,
		},
		{
			name:     "Configured custom provider",
			provider: "custom1",
			expected: true,
		},
		{
			name:     "Custom provider without API key",
			provider: "custom2",
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := config.IsProviderConfigured(tt.provider)
			if result != tt.expected {
				t.Errorf("Expected %v, got %v", tt.expected, result)
			}
		})
	}
}

func TestListConfiguredProviders(t *testing.T) {
	config := &Config{
		APIKeys: map[string]string{
			"openai":    "test-openai-key",
			"anthropic": "test-anthropic-key",
		},
		CustomProviders: map[string]CustomProvider{
			"custom1": {
				APIKey:  "custom1-key",
				BaseURL: "http://localhost:8080",
				Enabled: true,
			},
			"custom2": {
				BaseURL: "http://localhost:8081",
				Enabled: false, // Disabled
				APIKey:  "custom2-key",
			},
		},
	}

	providers := config.ListConfiguredProviders()

	// Should include openai, anthropic, ollama (local), and custom1
	// Should NOT include groq (no key) or custom2 (disabled)
	expectedProviders := map[string]bool{
		"openai":    true,
		"anthropic": true,
		"ollama":    true, // Local provider
		"custom1":   true,
	}

	if len(providers) != len(expectedProviders) {
		t.Errorf("Expected %d providers, got %d: %v", len(expectedProviders), len(providers), providers)
	}

	for _, provider := range providers {
		if !expectedProviders[provider] {
			t.Errorf("Unexpected provider: %s", provider)
		}
	}

	for expected := range expectedProviders {
		found := false
		for _, provider := range providers {
			if provider == expected {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("Expected provider not found: %s", expected)
		}
	}
}

File: typescript/tests/config.test.ts
// Mock fs-extra before importing ConfigManager
const mockPathExists = jest.fn();
const mockReadFile = jest.fn();
const mockWriteFile = jest.fn();
const mockEnsureDir = jest.fn();

jest.mock('fs-extra', () => ({
  pathExists: mockPathExists,
  readFile: mockReadFile,
  writeFile: mockWriteFile,
  ensureDir: mockEnsureDir
}));

import { ConfigManager } from '../src/lib/config';

const mockFs = {
  pathExists: mockPathExists,
  readFile: mockReadFile,
  writeFile: mockWriteFile,
  ensureDir: mockEnsureDir
};

describe('ConfigManager', () => {
  let configManager: ConfigManager;

  beforeEach(() => {
    jest.clearAllMocks();
    // Reset singleton
    (ConfigManager as any).instance = null;
    configManager = ConfigManager.getInstance();

    // Mock environment variables
    process.env.OPENAI_API_KEY = 'test-openai-key';
    process.env.ANTHROPIC_API_KEY = 'test-anthropic-key';
    process.env.PROMPD_DEFAULT_PROVIDER = 'openai';
    process.env.PROMPD_DEFAULT_MODEL = 'gpt-4';
  });

  afterEach(() => {
    // Clean up environment variables
    delete process.env.OPENAI_API_KEY;
    delete process.env.ANTHROPIC_API_KEY;
    delete process.env.PROMPD_DEFAULT_PROVIDER;
    delete process.env.PROMPD_DEFAULT_MODEL;
  });

  describe('loadConfig', () => {
    it('should load config with environment variables', async () => {
      mockFs.pathExists.mockResolvedValue(false);

      const config = await configManager.loadConfig();

      expect(config.apiKeys.openai).toBe('test-openai-key');
      expect(config.apiKeys.anthropic).toBe('test-anthropic-key');
      expect(config.defaultProvider).toBe('openai');
      expect(config.defaultModel).toBe('gpt-4');
    });

    it('should load config from file when available', async () => {
      const configContent = `
apiKeys:
  openai: file-openai-key
customProviders:
  local:
    baseUrl: http://localhost:8080
    models: [model1, model2]
    enabled: true
`;

      mockFs.pathExists.mockImplementation(async (path: string) => {
        return path.includes('config.yaml');
      });
      mockFs.readFile.mockResolvedValue(configContent);

      const config = await configManager.loadConfig();

      expect(config.apiKeys.openai).toBe('test-openai-key'); // Env overrides file
      expect(config.customProviders.local).toBeDefined();
      expect(config.customProviders.local.baseUrl).toBe('http://localhost:8080');
    });

    it('should handle missing config files gracefully', async () => {
      mockFs.pathExists.mockResolvedValue(false);

      const config = await configManager.loadConfig();

      expect(config).toBeDefined();
      expect(config.apiKeys).toBeDefined();
      expect(config.customProviders).toBeDefined();
    });
  });

  describe('getApiKey', () => {
    it('should return API key from config', async () => {
      mockFs.pathExists.mockResolvedValue(false);
      const config = await configManager.loadConfig();

      const apiKey = configManager.getApiKey('openai', config);
      expect(apiKey).toBe('test-openai-key');
    });

    it('should return undefined for unknown provider', async () => {
      mockFs.pathExists.mockResolvedValue(false);
      const config = await configManager.loadConfig();

      const apiKey = configManager.getApiKey('unknown-provider', config);
      expect(apiKey).toBeUndefined();
    });
  });

  describe('isProviderConfigured', () => {
    it('should return true for configured providers', async () => {
      mockFs.pathExists.mockResolvedValue(false);
      const config = await configManager.loadConfig();

      expect(configManager.isProviderConfigured('openai', config)).toBe(true);
      expect(configManager.isProviderConfigured('ollama', config)).toBe(true); // Local provider
    });

    it('should return false for unconfigured providers', async () => {
      delete process.env.OPENAI_API_KEY;
      delete process.env.ANTHROPIC_API_KEY;
      // Reset singleton to force reload
      (ConfigManager as any).instance = null;
      configManager = ConfigManager.getInstance();
      mockFs.pathExists.mockResolvedValue(false);
      const config = await configManager.loadConfig();

      expect(configManager.isProviderConfigured('openai', config)).toBe(false);
    });
  });

  describe('addCustomProvider', () => {
    it('should add custom provider and save config', async () => {
      mockFs.pathExists.mockResolvedValue(false);
      mockFs.ensureDir.mockResolvedValue(undefined);
      mockFs.writeFile.mockResolvedValue(undefined);

      await configManager.addCustomProvider(
        'test-provider',
        'http://localhost:8080',
        ['model1', 'model2'],
        'test-api-key'
      );

      expect(mockFs.writeFile).toHaveBeenCalled();
    });
  });

  describe('removeCustomProvider', () => {
    it('should remove existing custom provider', async () => {
      // First load a config with a custom provider
      const configContent = `
customProviders:
  test-provider:
    baseUrl: http://localhost:8080
    models: [model1]
    enabled: true
`;
      mockFs.pathExists.mockResolvedValue(true);
      mockFs.readFile.mockResolvedValue(configContent);
      mockFs.ensureDir.mockResolvedValue(undefined);
      mockFs.writeFile.mockResolvedValue(undefined);

      await configManager.removeCustomProvider('test-provider');

      expect(mockFs.writeFile).toHaveBeenCalled();
    });

    it('should throw error for non-existent provider', async () => {
      mockFs.pathExists.mockResolvedValue(false);

      await expect(configManager.removeCustomProvider('non-existent')).rejects.toThrow("Provider 'non-existent' not found");
    });
  });
});
